Model/Indexer/Invalidate.php

```diff
@@ -0,0 +1,177 @@
+
+    /** @var NostoServiceIndex */
+    private $nostoServiceIndex;
+
+    /** @var ProductCollectionFactory */
+    private $productCollectionFactory;
+
+    /** @var NostoHelperAccount */
+    private $nostoHelperAccount;
+
+    /** @var InvalidateModeSwitcher */
+    private $modeSwitcher;
+
+    /**
+     * Invalidate constructor.
+     * @param NostoHelperAccount $nostoHelperAccount
+     * @param NostoHelperScope $nostoHelperScope
+     * @param NostoServiceIndex $nostoServiceIndex
+     * @param NostoLogger $logger
+     * @param ProductCollectionFactory $productCollectionFactory
+     * @param InvalidateModeSwitcher $modeSwitcher
+     * @param StoreDimensionProvider $dimensionProvider
+     * @param Emulation $storeEmulation
+     * @param ProcessManager $processManager
+     * @param InputInterface $input
+     * @param Mview $mview
+     */
+    public function __construct(
+        NostoHelperAccount $nostoHelperAccount,
+        NostoHelperScope $nostoHelperScope,
+        NostoServiceIndex $nostoServiceIndex,
+        NostoLogger $logger,
+        ProductCollectionFactory $productCollectionFactory,
+        InvalidateModeSwitcher $modeSwitcher,
+        StoreDimensionProvider $dimensionProvider,
+        Emulation $storeEmulation,
+        ProcessManager $processManager,
+        InputInterface $input,
+        Mview $mview
+    ) {
+        $this->nostoServiceIndex = $nostoServiceIndex;
+        $this->nostoHelperAccount = $nostoHelperAccount;
+        $this->productCollectionFactory = $productCollectionFactory;
+        $this->modeSwitcher = $modeSwitcher;
+        parent::__construct(
+            $nostoHelperAccount,
+            $nostoHelperScope,
+            $logger,
+            $dimensionProvider,
+            $storeEmulation,
+            $input,
+            $mview,
+            $processManager
+        );
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getModeSwitcher(): ModeSwitcherInterface
+    {
+        return $this->modeSwitcher;
+    }
+
+    /**
+     * @inheritdoc
+     * @throws NostoException
+     */
+    public function doIndex(Store $store, array $ids = [])
+    {
+        if (!$this->nostoHelperAccount->nostoInstalledAndEnabled($store)) {
+            return;
+        }
+        $productCollection = $this->getCollection($store, $ids);
+        $this->nostoServiceIndex->invalidateOrCreate($productCollection, $store);
+
+        if (!empty($ids)) {
+            //In case for this specific set of ids
+            //there are more entries of products in the indexer table than the magento product collection
+            //it means that some products were deleted
+            $ids = array_unique($ids);
+            $idsSize = count($ids);
+            $collectionSize = $productCollection->getSize();
+            if ($idsSize > $collectionSize) {
+                $this->nostoServiceIndex->markProductsAsDeletedByDiff($productCollection, $ids, $store);
```

Comment:
Is the collection being iterated twice?

Reply:
Yes, but I think it would be a  micro optimization. This only happens happens when products are deleted and the max amount of products to be handled in one go is 1000.